compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2481,6 +2481,15 @@ impl<'a> Parser<'a> {
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr.kind, ExprKind::Block(..))
+            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && self.may_recover()
+        {
+            self.sess.emit_err(MissingExpressionInForLoop { span: expr.span });
```

Comment:
I wouldn't use `expr.span` here -- maybe use `expr.span.shrink_to_lo()` or something?

That's also the span you can use for the `/* expression */` suggestion that esteban gave.